### PR TITLE
Minimap: gamma-correct average texture colour calculation

### PR DESCRIPTION
**Pull request by HybridDog**
- Goal of the PR

Calculate the average texture colour while heeding the sRGB colourspace.

- How does the PR work?

#808080 sRGB is a lot darker than 0.5 gray, this should not be ignored when calculating an average colour for a texture.

- Does it resolve any reported issue?

No

- If not a bug fix, why is this PR needed? What usecases does it solve?

I have tested the minimap with a chessboard texture, and noticed that the average colour is wrong.
chessboard_pic
Before the changes:
current_minetest
After the changes:
gammacorrect_average_color

Note that browsers do incorrect image scaling, so the images must be viewed at 1:1 resolution or with nearest neighbour scaling.

‪

I also tried gamma-correct shading, but the results did look bad, and correct mini map shading is not important:
linear_phong_shading

## To do

This PR is a Ready for Review.

## How to test

Try textures with high contrast, such as chessboard.

### CODE REVIEW

File: src/client/tile.cpp [hunk 1]
```diff
@@ -2203,6 +2203,44 @@ video::ITexture* TextureSource::getNormalTexture(const std::string &name)
...
+	{
+		v3f col(col_srgb.X, col_srgb.Y, col_srgb.Z);
+		col *= 1.0f / 255.0f;
+		col.X = srgb_to_linear_component(col.X);
+		col.Y = srgb_to_linear_component(col.Y);
+		col.Z = srgb_to_linear_component(col.Z);
+		return col;
+	}
+
+	v3u8 linear_to_srgb(v3f &col_linear)
```
Comment (nerzhul) [inline src/client/tile.cpp:2230]: missing constness

File: src/client/tile.cpp [hunk 2]
```diff
@@ -2203,6 +2203,44 @@ video::ITexture* TextureSource::getNormalTexture(const std::string &name)
...
+		return 12.92f * v;
+	}
+	float srgb_to_linear_component(float v)
+	{
+		if (v > 0.04045f)
+			return powf((v + 0.055f) / 1.055f, 2.4f);
+		return v / 12.92f;
+	}
+
+	v3f srgb_to_linear(v3u8 col_srgb)
```
Comment (sfan5) [inline src/client/tile.cpp:2220]: Should this be `SColor` instead of `v3u8` (which we do not generally use)?

Comment (HybridDog) [reply src/client/tile.cpp]: It probably should. I don't remember why I have used v3u8.

File: src/irr_v3d.h [hunk 3]
```diff
@@ -27,4 +27,5 @@ typedef core::vector3df v3f;
 typedef core::vector3d<double> v3d;
 typedef core::vector3d<s16> v3s16;
 typedef core::vector3d<u16> v3u16;
+typedef core::vector3d<u8> v3u8;
```
Comment (sfan5) [inline src/irr_v3d.h:30]: please remove

File: src/client/tile.cpp [hunk 4]
```diff
@@ -2225,17 +2262,14 @@ video::SColor TextureSource::getTextureAverageColor(const std::string &name)
...
+				col_acc += srgb_to_linear(c);
 			}
 		}
 	}
 	image->drop();
 	if (total > 0) {
-		c.setRed(tR / total);
-		c.setGreen(tG / total);
-		c.setBlue(tB / total);
+		col_acc *= 1.0f / total;
```
Comment (sfan5) [inline src/client/tile.cpp:2271]: just use `/= total`?

File: src/client/tile.cpp [hunk 5]
```diff
@@ -2204,6 +2204,45 @@ video::ITexture* TextureSource::getNormalTexture(const std::string &name)
 	return NULL;
 }
 
+namespace {
+	float linear_to_srgb_component(const float v)
```
Comment (sfan5) [inline src/client/tile.cpp:2208]: please mark this and the below func as inline (also you don't need the `const float`)

would also be nice to have a comment here where these values come from

File: src/client/tile.cpp [hunk 6]
```diff
@@ -2204,6 +2204,45 @@ video::ITexture* TextureSource::getNormalTexture(const std::string &name)
...
+	{
+		if (v > 0.04045f)
+			return powf((v + 0.055f) / 1.055f, 2.4f);
+		return v / 12.92f;
+	}
+
+	v3f srgb_to_linear(const video::SColor &col_srgb)
+	{
+		v3f col(col_srgb.getRed(), col_srgb.getGreen(), col_srgb.getBlue());
+		col *= 1.0f / 255.0f;
```
Comment (sfan5) [inline src/client/tile.cpp:2224]: same with `/=` here

**Review by sfan5 — APPROVED** [review]

**Comment by rubenwardy** [comment]
This works well.

However, I'm unsure about the use of sRGB, do you have any resources to read on this? I've always average colours by squaring the components, getting the mean, and then averaging like so https://sighack.com/post/averaging-rgb-colors-the-right-way

**Comment by HybridDog** [comment]
Yes, here are some links:
Explanation of sRGB, physical and perceptual colour: https://bottosson.github.io/posts/colorwrong/
Gamma-correction for OpenGL and impact on shading: https://learnopengl.com/Advanced-Lighting/Gamma-Correction
Impact of gamma-correction on image scaling: www.ericbrasseur.org/gamma.html

**Review by SmallJoker — APPROVED** [review]
Colors look accurate.

**Pull request merged**